src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -34,7 +34,32 @@ public async Task RunAsync(ActionRunStage stage)
             Trace.Entering();
             ArgUtil.NotNull(ExecutionContext, nameof(ExecutionContext));
             ArgUtil.NotNull(Inputs, nameof(Inputs));
-            ArgUtil.NotNull(Data.Steps, nameof(Data.Steps));
+
+            List<Pipelines.ActionStep> steps;
+
+            if (stage == ActionRunStage.Pre)
+            {
+                ArgUtil.NotNull(Data.PreSteps, nameof(Data.PreSteps));
```

Comment:
discussed offline. follow-up to make this level-1 steps, not flat list. need to recurse (IActionRunner.cs) for Output procressing